adds missing formats in the registry (#3167)
OAI/OpenAPI-Specification: registries/_format/base64url.md

---
owner: baywet
issue:

> description: A url-safe binary array as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5)

Reviewer: I'm a bit confused at the combination of "binary array", which sounds like the datatype you are targeting, along with "url-safe", which refers to the encoded text.  I think what you want here is:

> Binary data encoded as a url-safe string as defined...

I would not mention the target data type at all, as the appropriate data type will vary among languages.  For example, I'm not sure what I'd do with "binary array" in Python.  In my view, the OAS should trust implementers to pick the right way to represent binary data, and avoid expressing a preference.

Author: ```suggestion
description: Binary data encoded as a url-safe string as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5)
```

base_type: string
layout: default
oas_version: < 3.1

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use [`contentEncoding` with a value of `base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding)."

Reviewer: I'm going to suggest a bit of streamlining here and elsewhere after looking at the rendered page.  Do feel free to ignore!  I do not feel strongly about this.

> In OpenAPI 3.1, instead use [`contentEncoding: base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding), optionally alongside [`contentMediaType`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentmediatype)

I wouldn't say "or above" because it's generally implied and if we do ship a 3.2 the recommendation could change anyway.

Author: ```suggestion
remarks: "When using OpenAPI 3.1 it's recommended not to use this format and instead use [`contentEncoding` with a value of `base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding)."
```

---

# <a href="..">{{ page.collection }}</a>

## {{ page.slug }} - {{ page.description }}

Base type: `{{ page.base_type }}`.

> The `{{page.slug}}` format is a url-safe binary array as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5).

Reviewer: I'd carry over the wording change away from "binary array" as noted above.

Author: ```suggestion
The `{{page.slug}}` format is binary data encoded as a url-safe string as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5).
```

### Upgrading from < 3.1 to >= 3.1

> #### Entire response body

Reviewer: I might add a note here to the effect of:

> Encoding a binary message body to a string is not necessary in HTTP, and will increase the size by 33% to 300%.

Author: added

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema:
                type: string

>                 format: byte

Reviewer: This should be `format: base64url`.

Author: ```suggestion
                format: base64url
```

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              type: string

>               contentEncoding: base64

Reviewer: This should be `contentEncoding: base64url`

Author: ```suggestion
              contentEncoding: base64url
```

```

#### Encoded property as part of a structured response

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            application/json:
              schema:
                properties:
                  width:
                    schema:
                      type: string
                  height:
                    schema:
                      type: string
                  image:
                    type: string
                    format: base64url

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            application/json:
              schema:
                properties:
                  width:
                    schema:
                      type: string
                  height:
                    schema:
                      type: string
                  image:
                    type: string
                    contentMediaType: image/png
                    contentEncoding: base64url
```

#### Query or Path parameters

```yaml
#OAS 3.0
parameters:
  - name: anImage
    in: query
    content:
      image/png:
        schema:
          type: string

>           format: byte 

Reviewer: `format: base64url`

Author: ```suggestion
          format: base64url
```

#OAS 3.1
parameters:
  - name: anImage
    in: query
    schema:
      type: string
      contentMediaType: image/png
      contentEncoding: base64url
```

#### Multipart/form-data

> Using base64url for multipart/form-data parts is not recommended since it artificially inflates the payload size with not added value.

Reviewer: I hadn't read this when I wrote the above wording, but adapting my wording would produce:

> Encoding a binary part to a string is not necessary for `multipart/form-data`, and will increase the size by 33% to 300%.

Author: updated

---
owner: baywet
issue:

> description: A url-safe binary array as defined in RFC4648

Reviewer: Is this being added purely for OAS 3.0?  In OAS 3.1 this should be `"contentEncoding": "base64url"`.

Author: I guess the format registry shows yup with the latest spec so we should design this documentation with 3.1 in mind. That being said OpenApi.net doesn't support 3.1 just yet (we're working on it) and kiota as of today looks at things from a 3.0 perspective to provide some context.
Do you think we should add a note here saying that with 3.1 people should refrain from using this and use the content encoding instead?

Reviewer: Yes, because the point of aligning OAS 3.1 and JSON Schema 2020-12 was to enable using off-the-shelf JSON Schema 2020-12 tools as much as possible. And those should know `"contentEncoding": "base64url"` but likely will not know `"format": "base64url"` (if they even pay attention to `format` at all, which many tools do not).

Author: noted, waiting for a resolution on the binary discussion before I push a change.

Author: added

---
owner: baywet
issue:
description: A url-safe binary array as defined in RFC4648
base_type: string

> layout: default

Reviewer 2: ```suggestion
layout: default
oas_version: < 3.1
remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use [`contentEncoding` with a value of `base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding)."
```

Author: for this one I'm not sure the markdown in a yaml field will be parsed properly by the templating engine. Would you mind if we keep it to plain text?

Reviewer 2: Not at all

Author: ```suggestion
layout: default
oas_version: < 3.1
remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use contentEncoding with a value of base64url."
```

Author: actually it looks like MD works, let me revert to what you originally suggested

---

# <a href="..">{{ page.collection }}</a>

## {{ page.slug }} - {{ page.description }}

Base type: `{{ page.base_type }}`.

> The `{{page.slug}}` format is a url-safe binary array as defined in RFC4648.

Reviewer 2: ```suggestion
The `{{page.slug}}` format is a url-safe binary array as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5).
```

---
owner: baywet
issue:
description: A url-safe binary array as defined in RFC4648
base_type: string
layout: default
oas_version: < 3.1

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use contentEncoding with a value of base64url."

Author: ```suggestion
remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use [`contentEncoding` with a value of `base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding)."
```

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use contentEncoding with a value of base64url."

Author: waiting for the resolution of this thread before we do so.
https://github.com/OAI/OpenAPI-Specification/pull/3167#discussion_r1113067249
And it looks like what we've just added to base64url might need some rewording if we decide using it for properties of the payload is ok in 3.1

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead use [`contentEncoding` with a value of `base64url`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-contentencoding)."

Reviewer 2: Cool! Can we have similar text in `byte.md`, with "base64" instead of "base64url"?

Can't propose it there because it's in the "old" part of the file.

Author: waiting for the resolution of this thread before we do so.
https://github.com/OAI/OpenAPI-Specification/pull/3167#discussion_r1113067249
And it looks like what we've just added to base64url might need some rewording if we decide using it for properties of the payload is ok in 3.1

---
owner: baywet
issue:

> description: A url-safe binary array as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5)

Reviewer 2: How about
```suggestion
description: base64url encoded data as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5)
```

> The `{{page.slug}}` format is a url-safe binary array as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5).

Reviewer 2: ```suggestion
The `{{page.slug}}` format represents base64url encoded data as defined in [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-5).
```

### Upgrading from < 3.1 to >= 3.1

#### Entire response body

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema:
                type: string

>                 format: byte

Reviewer 2: I think this example doesn't make sense. `"content":{"image/png":...}` means that the response will have `Content-Type: image/png`, and the response body should be the raw octet stream of a PNG image.

A correct example would be

~~~yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema:
                type: string
                format: binary
~~~

And it would belong to file `binary.md`.

A "correct" entire response body example for `format: base64url` would be with `Content-Type: text/plain`:

~~~yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            text/plain:
              schema:
                type: string
                format: base64url
~~~

However I am at a loss how the API would tell the client what is inside the base64url text, that would probably require out-of-band information, for example in a `description` field.

Reviewer: IIRC, when we were working through this for OAS 3.1, @darrelmiller said that in this case `Content-Type` would still be `image/png` because it is about the unencoded payload.  I might be misremembering _who_ argued for that, but it was someone because I remember having to change it 

The `contentEncoding` variation on this example is the [exact example](https://spec.openapis.org/oas/v3.1.0#considerations-for-file-uploads) in the OAS 3.1 spec, btw.  It comes with the following paragraph:

> Note that the `Content-Type` remains `image/png`, describing the semantics of the payload. The JSON Schema `type` and `contentEncoding` fields explain that the payload is transferred as text. The JSON Schema `contentMediaType` is technically redundant, but can be used by JSON Schema tools that may not be aware of the OpenAPI context.

I agree that it is a very confusing case, and as noted elsewhere in this discussion, there's not much reason to use it at all anyway.

The [corresponding example](https://spec.openapis.org/oas/v3.0.3#considerations-for-file-uploads) in OAS 3.0 is more ambiguous (note that that spec is confused over `byte` vs `base64` for the correct value to use with `format`).

Note that there was never a way to indicate the encoded media type in OAS 3.0 unless it was in the key for the Media Type Object or a scenario where an Encoding Object was used.

Author: Should we merge that as is and revisit when https://github.com/OAI/OpenAPI-Specification/issues/3174 is adressed?

Reviewer: I kind of feel like we should omit the examples that don't make sense (even if they are currently present in the spec).  Let me think a bit- I'm going to write a summary comment about finishing this PR shortly.

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              type: string
              contentEncoding: base64
```

> #### Encoded property as part of a structured response

Reviewer 2: ```suggestion
#### Encoded property as part of a JSON response
```

Author: same comment about structured types using other formats

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            application/json:
              schema:

>                 properties:

Reviewer 2: ```suggestion
                type: object
                properties:
```

                  width:
                    schema:
                      type: string
                  height:
                    schema:
                      type: string
                  image:
                    type: string
                    format: base64url

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            application/json:
              schema:

>                 properties:

Reviewer 2: ```suggestion
                type: object
                properties:
```
